### ISSUE TITLE
Return value of map shared when referencing a sub in CORE::

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12127]
Migrated from [rt.perl.org#113044](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113044) (status was 'resolved')

Searchable as RT113044$

**Comment by p5pRT** [comment]
### From gustavderdrache@gmail.com

#### Created by gustavderdrache@gmail.com



When using a subref in CORE&#8203;:: \(e\.g\.\, \\&CORE&#8203;::oct\)\, the return value is shared across all iterations of map\.  This can be alleviated by using an expression such as 0\+$foo\.

The script below reproduces the issue&#8203;:

&nbsp;    $oct = \\&CORE&#8203;::oct;
&nbsp;    
&nbsp;    @&#8203;mapped = map &$oct\($\_\)\, qw\(0b10 010 0x10\);
&nbsp;    print "@&#8203;mapped\\n";
&nbsp;    
&nbsp;    @&#8203;mapped = map 0\+&$oct\($\_\)\, qw\(0b10 010 0x10\);
&nbsp;    print "@&#8203;mapped\\n";
&nbsp;    

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.16.0:

Configured by alex at Tue May 22 11:25:51 EDT 2012.

Summary of my perl5 (revision 5 version 16 subversion 0) configuration:
   
  Platform:
    osname=darwin, osvers=11.4.0, archname=darwin-thread-multi-2level
    uname='darwin ghost-bear.local 11.4.0 darwin kernel version 11.4.0: mon apr 9 19:32:15 pdt 2012; root:xnu-1699.26.8~1release_x86_64 x86_64 '
    config_args='-de -Dprefix=/Users/alex/perl5/perlbrew/perls/perl-5.16.0 -Duseithreads -Duse64bitint'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple Clang 3.1 (tags/Apple/clang-318.0.45)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib
    libs=-ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.16.0:
    /Users/alex/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0/darwin-thread-multi-2level
    /Users/alex/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0
    /Users/alex/perl5/perlbrew/perls/perl-5.16.0/lib/5.16.0/darwin-thread-multi-2level
    /Users/alex/perl5/perlbrew/perls/perl-5.16.0/lib/5.16.0
    .


Environment for perl 5.16.0:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/alex
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/alex/Library/Haskell/bin:/Users/alex/perl5/perlbrew/bin:/Users/alex/perl5/perlbrew/perls/perl-5.16.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/X11/bin:/Users/alex/bin
    PERLBREW_BASHRC_VERSION=0.33
    PERLBREW_HOME=/Users/alex/.perlbrew
    PERLBREW_PATH=/Users/alex/perl5/perlbrew/bin:/Users/alex/perl5/perlbrew/perls/perl-5.16.0/bin
    PERLBREW_PERL=perl-5.16.0
    PERLBREW_ROOT=/Users/alex/perl5/perlbrew
    PERLBREW_VERSION=0.33
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 22 13&#8203;:21&#8203;:35 2012\, gustavderdrache@&#8203;gmail\.com wrote&#8203;:
> This is a bug report for perl from gustavderdrache@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> When using a subref in CORE&#8203;:: \(e\.g\.\, \\&CORE&#8203;::oct\)\, the return value is
> shared across all iterations of map\.  This can be alleviated by using
> an expression such as 0\+$foo\.
> 
> The script below reproduces the issue&#8203;:
> 
>     $oct = \\&CORE&#8203;::oct;
> 
>     @&#8203;mapped = map &$oct\($\_\)\, qw\(0b10 010 0x10\);
>     print "@&#8203;mapped\\n";
> 
>     @&#8203;mapped = map 0\+&$oct\($\_\)\, qw\(0b10 010 0x10\);
>     print "@&#8203;mapped\\n";

Yes\, I discovered this recently\.  The same issue affects lvalue subs\.  I
fixed it in commit 5811c07eab3\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 22 13&#8203;:39&#8203;:32 2012\, sprout wrote&#8203;:
> On Tue May 22 13&#8203;:21&#8203;:35 2012\, gustavderdrache@&#8203;gmail\.com wrote&#8203;:
> > This is a bug report for perl from gustavderdrache@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > When using a subref in CORE&#8203;:: \(e\.g\.\, \\&CORE&#8203;::oct\)\, the return value is
> > shared across all iterations of map\.  This can be alleviated by using
> > an expression such as 0\+$foo\.
> > 
> > The script below reproduces the issue&#8203;:
> > 
> >     $oct = \\&CORE&#8203;::oct;
> > 
> >     @&#8203;mapped = map &$oct\($\_\)\, qw\(0b10 010 0x10\);
> >     print "@&#8203;mapped\\n";
> > 
> >     @&#8203;mapped = map 0\+&$oct\($\_\)\, qw\(0b10 010 0x10\);
> >     print "@&#8203;mapped\\n";
> 
> Yes\, I discovered this recently\.  The same issue affects lvalue subs\.  I
> fixed it in commit 5811c07eab3\.
> 

And it looks as though the fix will be in 5\.16\.1\.

\-\- 

Father Chrysostomos
